README.md

<img src="https://github.com/GarrettJenkinson/condor_pytorch/blob/main/docs/img/condor.png" width=300>

**CONDOR pytorch implementation for ordinal regression with deep neural networks.**

![Continuous Integration](https://github.com/GarrettJenkinson/condor_pytorch/actions/workflows/python-test.yml/badge.svg)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/GarrettJenkinson/condor_pytorch/blob/master/LICENSE)
![Python 3](https://img.shields.io/badge/python-3-blue.svg)

<br>

**Documentation: [https://GarrettJenkinson.github.io/condor_pytorch](https://GarrettJenkinson.github.io/condor_pytorch)**

---

## About  

CONDOR, short for CONDitionals for Ordinal Regression, is a method for ordinal regression with deep neural networks, 
which addresses the rank inconsistency issue of other ordinal regression frameworks.

It is compatible with any state-of-the-art deep neural network architecture, requiring only modification of the output layer, the labels, the loss function.

This repository implements the CONDOR functionality (neural network layer, loss function, and dataset utilities) for convenient use. 
Examples are provided via the "Tutorials" that can be found on the documentation website at [https://GarrettJenkinson.github.io/condor_pytorch](https://GarrettJenkinson.github.io/condor_pytorch).

We also have [CONDOR implemented for Tensorflow](https://github.com/GarrettJenkinson/condor_tensorflow).

<img src="https://github.com/GarrettJenkinson/condor_pytorch/raw/main/docs/img/rankconsistent.png" width=500>

---

## Installation or Docker
---

You can install the latest stable release of `condor_pytorch` directly from Python's package index via `pip` by executing the following code from your command line:  

```bash
pip install condor-pytorch
```

The dependencies can also be pip installed also using the included `requirements.txt`:

```bash
pip install -r requirements.txt
```

We also provide Dockerfile's to help get up and started quickly with `condor_pytorch`.
The cpu image can be built and ran as follows, with tutorial jupyter notebooks
built in.

```bash
# Create a docker image, only done once
docker build -t cpu_pytorch -f cpu.Dockerfile ./

# run image to serve a jupyter notebook
docker run -it -p 8888:8888 --rm cpu_pytorch

# how to run bash inside container (with python that will have deps)
docker run -u $(id -u):$(id -g) -it -p 8888:8888 --rm cpu_pytorch bash
```

An NVIDIA based gpu optimized container can be built and run 
as follows (without interactive ipynb capabilities).

```bash
# only needs to be built once
docker build -t gpu_pytorch -f gpu.Dockerfile ./

# use the image after building it
docker run -it -p 8888:8888 --rm gpu_pytorch
```

## Cite as

If you use CONDOR as part of your workflow in a scientific publication, please consider citing the CONDOR repository with the following DOI:

- [Jenkinson, Khezeli, Oliver, Kalantari, Klee. Universally rank consistent ordinal regression in neural networks, arXiv:2110.07470, 2021.](https://arxiv.org/abs/2110.07470)

```
@article{condor2021,
title = "Universally rank consistent ordinal regression in neural networks",
journal = "arXiv",
volume = "2110.07470",
year = "2021",
url = "https://arxiv.org/abs/2110.07470",
author = "Garrett Jenkinson and Kia Khezeli and Gavin R. Oliver and John Kalantari and Eric W. Klee",
keywords = "Deep learning, Ordinal regression, neural networks, Machine learning, Biometrics"
}
```

**Acknowledgments**: Many thanks to [the CORAL ordinal authors](https://github.com/ck37/coral-ordinal) and 
[the CORAL pytorch authors](https://github.com/Raschka-research-group/coral_pytorch) whose repos provided a roadmap for this codebase.


docs/CHANGELOG.md
# Release Notes

The changelog for the current development version is available at
[https://github.com/GarrettJenkinson/condor_pytorch/blob/main/docs/CHANGELOG.md](https://github.com/GarrettJenkinson/condor_pytorch/blob/main/docs/CHANGELOG.md).

### 1.1.0 (02/19/2022)

##### Downloads

- [Source code (zip)](https://github.com/GarrettJenkinson/condor_pytorch/archive/v1.1.0.zip)
- [Source code (tar.gz)](https://github.com/GarrettJenkinson/condor_pytorch/archive/v1.1.0.tar.gz)

##### New Features

- The `condor_negloglikeloss` function has been implemented to provide the maximume likelihood estimates of network parameters.

##### Changes

- None.

##### Bug Fixes

- None.

### 1.0.1 (11/01/2021)

##### Downloads

- [Source code (zip)](https://github.com/GarrettJenkinson/condor_pytorch/archive/v1.0.1.zip)
- [Source code (tar.gz)](https://github.com/GarrettJenkinson/condor_pytorch/archive/v1.0.1.tar.gz)

##### New Features

- First release.

##### Changes

- First release.

##### Bug Fixes

- First release.


tests/test_losses.py
from condor_pytorch.losses import CondorOrdinalCrossEntropy
from condor_pytorch.losses import condor_negloglikeloss
import pytest
import torch


def test_CondorOrdinalCrossEntropy():
    levels = torch.tensor([[1., 1., 0., 0.],
                           [1., 0., 0., 0.],
                           [1., 1., 1., 1.]])
    logits = torch.tensor([[2.1, 1.8, -2.1, -1.8],
                           [1.9, -1., -1.5, -1.3],
                           [1.9, 1.8, 1.7, 1.6]])
    assert torch.allclose(CondorOrdinalCrossEntropy(logits,levels),
                          torch.tensor(0.8259),atol=1e-4,rtol=1e-4)

def test_CondorNLL():
    levels = torch.tensor([[1., 1., 0., 0.],
                           [1., 0., 0., 0.],
                           [1., 1., 1., 1.]])
    logits = torch.tensor([[2.1, 1.8, -2.1, -1.8],
                           [1.9, -1., -1.5, -1.3],
                           [1.9, 1.8, 1.7, 1.6]])
    assert torch.allclose(condor_negloglikeloss(logits,levels),
                          torch.tensor(0.4936),atol=1e-4,rtol=1e-4)


tests/test_dataset.py
from condor_pytorch.dataset import label_to_levels
from condor_pytorch.dataset import levels_from_labelbatch
from condor_pytorch.dataset import proba_to_label
from condor_pytorch.dataset import logits_to_label
import pytest
import torch


def test_label_to_levels():
    assert torch.allclose(label_to_levels(0, num_classes=5),
                          torch.tensor([0., 0., 0., 0.]))
    assert torch.allclose(label_to_levels(1, num_classes=5),
                          torch.tensor([1., 0., 0., 0.]))
    assert torch.allclose(label_to_levels(2, num_classes=5),
                          torch.tensor([1., 1., 0., 0.]))
    assert torch.allclose(label_to_levels(3, num_classes=5),
                          torch.tensor([1., 1., 1., 0.]))
    assert torch.allclose(label_to_levels(4, num_classes=5),
                          torch.tensor([1., 1., 1., 1.]))

def test_levels_from_labelbatch():
    assert torch.allclose(levels_from_labelbatch(labels=[2, 1, 4],
                                                 num_classes=5),
                          torch.tensor([[1., 1., 0., 0.],
                                        [1., 0., 0., 0.],
                                        [1., 1., 1., 1.]]))

def test_proba_to_label():
    probas = torch.tensor([[0.934, 0.861, 0.323, 0.492, 0.295],
                           [0.496, 0.485, 0.267, 0.124, 0.058],
                           [0.985, 0.967, 0.920, 0.819, 0.506]])
    assert torch.allclose(proba_to_label(probas),
                         torch.tensor([2, 0, 5]))

def test_logits_to_label():
    logits = torch.tensor([[ 0.934, -0.861,  0.323, -0.492, -0.295],
                           [-0.496,  0.485,  0.267,  0.124, -0.058],
                           [ 0.985,  0.967, -0.920,  0.819, -0.506]])
    assert torch.allclose(logits_to_label(logits),
                         torch.tensor([1, 0, 2]))



tests/test_metrics.py
from condor_pytorch.metrics import earth_movers_distance
from condor_pytorch.metrics import ordinal_accuracy
from condor_pytorch.metrics import mean_absolute_error
import pytest
import torch

def test_EMD():
    levels = torch.tensor(
       [[1., 1., 0., 0.],
        [1., 0., 0., 0.],
       [1., 1., 1., 1.]])
    logits = torch.tensor(
       [[2.1, 1.8, -2.1, -1.8],
        [1.9, -1., -1.5, -1.3],
        [1.9, 1.8, 1.7, 1.6]])

    assert torch.allclose(earth_movers_distance(logits,levels),
                          torch.tensor(0.6943),atol=1e-4,rtol=1e-4)

def test_ACC():
   levels = torch.tensor(
       [[1., 1., 0., 0.],
        [1., 0., 0., 0.],
        [1., 1., 1., 1.]])
   logits = torch.tensor(
      [[2.1, 1.8, -2.1, -1.8],
       [1.9, -1., -1.5, -1.3],
       [1.9, 1.8, 1.7, 1.6]])
   assert torch.allclose(ordinal_accuracy(logits, levels),
                         torch.tensor(1.),atol=1e-4,rtol=1e-4)

def test_MAE():
   levels = torch.tensor(
      [[1., 1., 0., 0.],
       [1., 0., 0., 0.],
      [1., 1., 1., 1.]])
   logits = torch.tensor(
      [[2.1, 1.8, -2.1, -1.8],
       [1.9, -1., -1.5, -1.3],
       [1.9, 1.8, 1.7, 1.6]])
   assert torch.allclose(mean_absolute_error(logits, levels),
                         torch.tensor(0.),atol=1e-4,rtol=1e-4)


src/condor_pytorch/losses.py
import torch.nn.functional as F
import torch

# Modifed from: https://github.com/tensorflow/tensorflow/blob/6dcd6fcea73ad613e78039bd1f696c35e63abb32/tensorflow/python/ops/nn_impl.py#L112-L148
def condor_negloglikeloss(logits, labels, reduction='mean'):
    """computes the negative log likelihood loss described in

    condor tbd.

    parameters
    ----------
    logits : torch.tensor, shape(num_examples, num_classes-1)
        outputs of the condor layer.

    labels : torch.tensor, shape(num_examples, num_classes-1)
        true labels represented as extended binary vectors
        (via `condor_pytorch.dataset.levels_from_labelbatch`).

    reduction : str or none (default='mean')
        if 'mean' or 'sum', returns the averaged or summed loss value across
        all data points (rows) in logits. if none, returns a vector of
        shape (num_examples,)

    returns
    ----------
        loss : torch.tensor
        a torch.tensor containing a single loss value (if `reduction='mean'` or '`sum'`)
        or a loss value for each data record (if `reduction=none`).

    examples
    ----------
    >>> import torch
    >>> labels = torch.tensor(
    ...    [[1., 1., 0., 0.],
    ...     [1., 0., 0., 0.],
    ...    [1., 1., 1., 1.]])
    >>> logits = torch.tensor(
    ...    [[2.1, 1.8, -2.1, -1.8],
    ...     [1.9, -1., -1.5, -1.3],
    ...     [1.9, 1.8, 1.7, 1.6]])
    >>> condor_negloglikeloss(logits, labels)
    tensor(0.4936)
    """
    if not logits.shape == labels.shape:
        raise ValueError("Please ensure that logits (%s) has the same shape as labels (%s). "
                         % (logits.shape, labels.shape))
    piLab = torch.cat([torch.ones((labels.shape[0],1)),labels[:,:-1]],dim=1)

    # The logistic loss formula from above is
    #   x - x * z + log(1 + exp(-x))
    # For x < 0, a more numerically stable formula is
    #   -x * z + log(1 + exp(x))
    # Note that these two expressions can be combined into the following:
    #   max(x, 0) - x * z + log(1 + exp(-abs(x)))
    # To allow computing gradients at zero, we define custom versions of max and
    # abs functions.
    zeros = torch.zeros_like(logits, dtype=logits.dtype)
    cond = (logits >= zeros)
    cond2 = (piLab > zeros)
    relu_logits = torch.where(cond, logits, zeros)
    neg_abs_logits = torch.where(cond, -logits, logits)
    temp = relu_logits - (logits*labels) + torch.log1p(torch.exp(neg_abs_logits))
    val = torch.sum(torch.where(cond2, temp, zeros),dim=1)

    if reduction == 'mean':
        loss = torch.mean(val)
    elif reduction == 'sum':
        loss = torch.sum(val)
    elif reduction is None:
        loss = val
    else:
        s = ('Invalid value for `reduction`. Should be "mean", '
             '"sum", or None. Got %s' % reduction)
        raise ValueError(s)

    return loss

def CondorOrdinalCrossEntropy(logits, levels, importance_weights=None, reduction='mean'):
    """computes the condor loss described in

    condor tbd.

    parameters
    ----------
    logits : torch.tensor, shape(num_examples, num_classes-1)
        outputs of the condor layer.

    levels : torch.tensor, shape(num_examples, num_classes-1)
        true labels represented as extended binary vectors
        (via `condor_pytorch.dataset.levels_from_labelbatch`).

    importance_weights : torch.tensor, shape=(num_classes-1,) (default=none)
        optional weights for the different labels in levels.
        a tensor of ones, i.e.,
        `torch.ones(num_classes-1, dtype=torch.float32)`
        will result in uniform weights that have the same effect as none.

    reduction : str or none (default='mean')
        if 'mean' or 'sum', returns the averaged or summed loss value across
        all data points (rows) in logits. if none, returns a vector of
        shape (num_examples,)

    returns
    ----------
        loss : torch.tensor
        a torch.tensor containing a single loss value (if `reduction='mean'` or '`sum'`)
        or a loss value for each data record (if `reduction=none`).

    examples
    ----------
    >>> import torch
    >>> levels = torch.tensor(
    ...    [[1., 1., 0., 0.],
    ...     [1., 0., 0., 0.],
    ...    [1., 1., 1., 1.]])
    >>> logits = torch.tensor(
    ...    [[2.1, 1.8, -2.1, -1.8],
    ...     [1.9, -1., -1.5, -1.3],
    ...     [1.9, 1.8, 1.7, 1.6]])
    >>> CondorOrdinalCrossEntropy(logits, levels)
    tensor(0.8259)
    """

    if not logits.shape == levels.shape:
        raise ValueError("Please ensure that logits (%s) has the same shape as levels (%s). "
                         % (logits.shape, levels.shape))

    logprobs = torch.cumsum(F.logsigmoid(logits), dim = 1)
    term1 = (logprobs*levels
             + torch.log(1 - torch.exp(logprobs)+torch.finfo(torch.float32).eps)*(1-levels))

    if importance_weights is not None:
        term1 *= importance_weights

    val = (-torch.sum(term1, dim=1))

    if reduction == 'mean':
        loss = torch.mean(val)
    elif reduction == 'sum':
        loss = torch.sum(val)
    elif reduction is None:
        loss = val
    else:
        s = ('Invalid value for `reduction`. Should be "mean", '
             '"sum", or None. Got %s' % reduction)
        raise ValueError(s)

    return loss


src/condor_pytorch/metrics.py
import torch.nn.functional as F
import torch
from .activations import ordinal_softmax

def earth_movers_distance(logits, levels, device='cpu', reduction='mean'):
    """Computes the Earth Movers Distance

    Parameters
    ----------
    logits : torch.tensor, shape(num_examples, num_classes-1)
        Outputs of the CONDOR layer.

    levels : torch.tensor, shape(num_examples, num_classes-1)
        True labels represented as extended binary vectors
        (via `condor_pytorch.dataset.levels_from_labelbatch`).
        
    device: 'cpu', 'cuda', or None (default='cpu')
        If GPUs are utilized, then the device should be passed accordingly. 

    reduction : str or None (default='mean')
        If 'mean' or 'sum', returns the averaged or summed loss value across
        all data points (rows) in logits. If None, returns a vector of
        shape (num_examples,)

    Returns
    ----------
        loss : torch.tensor
        A torch.tensor containing a single loss value (if `reduction='mean'` or '`sum'`)
        or a loss value for each data record (if `reduction=None`).

    Examples
    ----------
    >>> import torch
    >>> levels = torch.tensor(
    ...    [[1., 1., 0., 0.],
    ...     [1., 0., 0., 0.],
    ...    [1., 1., 1., 1.]])
    >>> logits = torch.tensor(
    ...    [[2.1, 1.8, -2.1, -1.8],
    ...     [1.9, -1., -1.5, -1.3],
    ...     [1.9, 1.8, 1.7, 1.6]])
    >>> earth_movers_distance(logits, levels)
    tensor(0.6943)
    """
    nclasses = logits.shape[1]+1
    nbatch = logits.shape[0]
    if not logits.shape == levels.shape:
        raise ValueError("Please ensure that logits (%s) has the same shape as levels (%s). "
                         % (logits.shape, levels.shape))

    cum_probs = ordinal_softmax(logits, device)
    y_true = torch.sum(levels,dim=1,keepdim=True,dtype=logits.dtype)

#     y_dist = torch.abs(torch.tile(y_true,(1,nclasses))
#                        -torch.tile(torch.arange(0,nclasses),(nbatch,1)))
    y_dist = torch.abs(y_true.repeat(1,nclasses) - torch.arange(0,nclasses).repeat(nbatch,1)).to(device)
    
    val = torch.sum(torch.mul(cum_probs,y_dist),1)

    if reduction == 'mean':
        loss = torch.mean(val)
    elif reduction == 'sum':
        loss = torch.sum(val)
    elif reduction is None:
        loss = val
    else:
        s = ('Invalid value for `reduction`. Should be "mean", '
             '"sum", or None. Got %s' % reduction)
        raise ValueError(s)

    return loss

def ordinal_accuracy(logits, levels, device='cpu', tolerance=0, reduction='mean'):
    """Computes the accuracy with a tolerance for ordinal error.

    Parameters
    ----------
    logits : torch.tensor, shape(num_examples, num_classes-1)
        Outputs of the CONDOR layer.

    levels : torch.tensor, shape(num_examples, num_classes-1)
        True labels represented as extended binary vectors
        (via `condor_pytorch.dataset.levels_from_labelbatch`).
    
    device: 'cpu', 'cuda', or None (default='cpu')
        If GPUs are utilized, then the device should be passed accordingly. 
    
    tolerance : integer
        Allowed error in the ordinal ranks that will count as
        a correct prediction.

    reduction : str or None (default='mean')
        If 'mean' or 'sum', returns the averaged or summed loss value across
        all data points (rows) in logits. If None, returns a vector of
        shape (num_examples,)

    Returns
    ----------
        loss : torch.tensor
        A torch.tensor containing a single loss value (if `reduction='mean'` or '`sum'`)
        or a loss value for each data record (if `reduction=None`).

    Examples
    ----------
    >>> import torch
    >>> levels = torch.tensor(
    ...    [[1., 1., 0., 0.],
    ...     [1., 0., 0., 0.],
    ...    [1., 1., 1., 1.]])
    >>> logits = torch.tensor(
    ...    [[2.1, 1.8, -2.1, -1.8],
    ...     [1.9, -1., -1.5, -1.3],
    ...     [1.9, 1.8, 1.7, 1.6]])
    >>> ordinal_accuracy(logits, levels)
    tensor(1.)
    """
    nclasses = logits.shape[1]+1
    nbatch = logits.shape[0]
    if not logits.shape == levels.shape:
        raise ValueError("Please ensure that logits (%s) has the same shape as levels (%s). "
                         % (logits.shape, levels.shape))

    y_true = torch.sum(levels,dim=1,keepdim=True,dtype=logits.dtype).to(device)

    y_est = torch.sum(torch.cumprod(torch.sigmoid(logits),dim=1)>0.5,dim=1,keepdim=True,dtype=logits.dtype).to(device)
    
    # 1 when correct and 0 else
    val = torch.le(torch.abs(y_true-y_est),tolerance).to(torch.float32)

    if reduction == 'mean':
        loss = torch.mean(val)
    elif reduction == 'sum':
        loss = torch.sum(val)
    elif reduction is None:
        loss = val
    else:
        s = ('Invalid value for `reduction`. Should be "mean", '
             '"sum", or None. Got %s' % reduction)
        raise ValueError(s)

    return loss

def mean_absolute_error(logits, levels, reduction='mean'):
    """Computes the mean absolute error of ordinal predictions.

    Parameters
    ----------
    logits : torch.tensor, shape(num_examples, num_classes-1)
        Outputs of the CONDOR layer.

    levels : torch.tensor, shape(num_examples, num_classes-1)
        True labels represented as extended binary vectors
        (via `condor_pytorch.dataset.levels_from_labelbatch`).

    reduction : str or None (default='mean')
        If 'mean' or 'sum', returns the averaged or summed loss value across
        all data points (rows) in logits. If None, returns a vector of
        shape (num_examples,)

    Returns
    ----------
        loss : torch.tensor
        A torch.tensor containing a single loss value (if `reduction='mean'` or '`sum'`)
        or a loss value for each data record (if `reduction=None`).

    Examples
    ----------
    >>> import torch
    >>> levels = torch.tensor(
    ...    [[1., 1., 0., 0.],
    ...     [1., 0., 0., 0.],
    ...    [1., 1., 1., 1.]])
    >>> logits = torch.tensor(
    ...    [[2.1, 1.8, -2.1, -1.8],
    ...     [1.9, -1., -1.5, -1.3],
    ...     [1.9, 1.8, 1.7, 1.6]])
    >>> mean_absolute_error(logits, levels)
    tensor(0.)
    """
    nclasses = logits.shape[1]+1
    nbatch = logits.shape[0]
    if not logits.shape == levels.shape:
        raise ValueError("Please ensure that logits (%s) has the same shape as levels (%s). "
                         % (logits.shape, levels.shape))

    y_true = torch.sum(levels,dim=1,keepdim=True,dtype=logits.dtype)

    y_est = torch.sum(torch.cumprod(torch.sigmoid(logits),dim=1)>0.5,dim=1,keepdim=True,dtype=logits.dtype)

    # 1 when correct and 0 else
    val = torch.abs(y_true-y_est)

    if reduction == 'mean':
        loss = torch.mean(val)
    elif reduction == 'sum':
        loss = torch.sum(val)
    elif reduction is None:
        loss = val
    else:
        s = ('Invalid value for `reduction`. Should be "mean", '
             '"sum", or None. Got %s' % reduction)
        raise ValueError(s)

    return loss


src/condor_pytorch/__init__.py
from .version import __version__

from .activations import ordinal_softmax
from .dataset import label_to_levels
from .dataset import levels_from_labelbatch
from .dataset import proba_to_label
from .dataset import logits_to_label
from .losses import CondorOrdinalCrossEntropy
from .losses import condor_negloglikeloss
from .metrics import earth_movers_distance
from .metrics import mean_absolute_error
from .metrics import ordinal_accuracy
#from .labelencoder import CondorOrdinalEncoder

# if somebody does "from somepackage import *", this is what they will
# be able to access:
__all__ = [
  'ordinal_softmax',
  'label_to_levels',
  'levels_from_labelbatch',
  'proba_to_label',
  'logits_to_label',
  'CondorOrdinalCrossEntropy',
  'condor_negloglikeloss',
  'mean_absolute_error',
  'ordinal_accuracy',
#  'CondorOrdinalEncoder',
]


src/condor_pytorch/activations.py
import torch
import torch.nn.functional as F

def ordinal_softmax(x, device='cpu'):
    """ Convert the ordinal logit output to label probabilities.

    Parameters
    ----------
    x: torch.Tensor, shape=(num_samples,num_classes-1)
        Logit output of the final Dense(num_classes-1) layer.
        
    device: 'cpu', 'cuda', or None (default='cpu')
        If GPUs are utilized, then the device should be passed accordingly. 

    Returns
    ----------
    probs_tensor: torch.Tensor, shape=(num_samples, num_classes)
        Probabilities of each class (columns) for each
        sample (rows).

    Examples
    ----------
    >>> ordinal_softmax(torch.tensor([[-1.,1],[-2,2]]))
    tensor([[0.7311, 0.0723, 0.1966],
            [0.8808, 0.0142, 0.1050]])
    """
    
    # Convert the ordinal logits into cumulative probabilities.
    
    
    log_probs = F.logsigmoid(x).to(device)
    cum_probs = torch.cat((torch.ones(x.shape[0],1,dtype=torch.float32).to(device),
                         torch.exp(torch.cumsum(log_probs, dim = 1)),
                         torch.zeros(x.shape[0],1,dtype=torch.float32).to(device)),
                        dim=1)    
    return cum_probs[:,0:-1] - cum_probs[:,1:]


src/condor_pytorch/dataset.py
import torch


def label_to_levels(label, num_classes, dtype=torch.float32):
    """Converts integer class label to extended binary label vector

    Parameters
    ----------
    label : int
        Class label to be converted into a extended
        binary vector. Should be smaller than num_classes-1.

    num_classes : int
        The number of class clabels in the dataset. Assumes
        class labels start at 0. Determines the size of the
        output vector.

    dtype : torch data type (default=torch.float32)
        Data type of the torch output vector for the
        extended binary labels.

    Returns
    ----------
    levels : torch.tensor, shape=(num_classes-1,)
        Extended binary label vector. Type is determined
        by the `dtype` parameter.

    Examples
    ----------
    >>> label_to_levels(0, num_classes=5)
    tensor([0., 0., 0., 0.])
    >>> label_to_levels(1, num_classes=5)
    tensor([1., 0., 0., 0.])
    >>> label_to_levels(3, num_classes=5)
    tensor([1., 1., 1., 0.])
    >>> label_to_levels(4, num_classes=5)
    tensor([1., 1., 1., 1.])
    """
    if not label <= num_classes-1:
        raise ValueError('Class label must be smaller or '
                         'equal to %d (num_classes-1). Got %d.'
                         % (num_classes-1, label))
    if isinstance(label, torch.Tensor):
        int_label = label.item()
    else:
        int_label = label

    levels = [1]*int_label + [0]*(num_classes - 1 - int_label)
    levels = torch.tensor(levels, dtype=dtype)
    return levels


def levels_from_labelbatch(labels, num_classes, dtype=torch.float32):
    """
    Converts a list of integer class label to extended binary label vectors

    Parameters
    ----------
    labels : list or 1D orch.tensor, shape=(num_labels,)
        A list or 1D torch.tensor with integer class labels
        to be converted into extended binary label vectors.

    num_classes : int
        The number of class clabels in the dataset. Assumes
        class labels start at 0. Determines the size of the
        output vector.

    dtype : torch data type (default=torch.float32)
        Data type of the torch output vector for the
        extended binary labels.

    Returns
    ----------
    levels : torch.tensor, shape=(num_labels, num_classes-1)

    Examples
    ----------
    >>> levels_from_labelbatch(labels=[2, 1, 4], num_classes=5)
    tensor([[1., 1., 0., 0.],
            [1., 0., 0., 0.],
            [1., 1., 1., 1.]])
    """
    levels = []
    for label in labels:
        levels_from_label = label_to_levels(
            label=label, num_classes=num_classes, dtype=dtype)
        levels.append(levels_from_label)

    levels = torch.stack(levels)
    return levels


def proba_to_label(probas):
    """
    Converts predicted probabilities from extended binary format
    to integer class labels

    Parameters
    ----------
    probas : torch.tensor, shape(n_examples, n_labels)
        Torch tensor consisting of probabilities returned by CORAL model.

    Examples
    ----------
    >>> # 3 training examples, 6 classes
    >>> probas = torch.tensor([[0.934, 0.861, 0.323, 0.492, 0.295],
    ...                        [0.496, 0.485, 0.267, 0.124, 0.058],
    ...                        [0.985, 0.967, 0.920, 0.819, 0.506]])
    >>> proba_to_label(probas)
    tensor([2, 0, 5])
    """
    predict_levels = probas > 0.5
    predicted_labels = torch.sum(predict_levels, dim=1)
    return predicted_labels


def logits_to_label(logits):
    """
    Converts predicted logits from extended binary format
    to integer class labels

    Parameters
    ----------
    logits : torch.tensor, shape(n_examples, n_labels-1)
        Torch tensor consisting of probabilities returned by ORCA model.

    Examples
    ----------
    >>> # 3 training examples, 6 classes
    >>> logits = torch.tensor([[ 0.934, -0.861,  0.323, -0.492, -0.295],
    ...                        [-0.496,  0.485,  0.267,  0.124, -0.058],
    ...                        [ 0.985,  0.967, -0.920,  0.819, -0.506]])
    >>> logits_to_label(logits)
    tensor([1, 0, 2])
    """
    probas = torch.cumprod(torch.sigmoid(logits), dim=1)
    predict_levels = probas > 0.5
    predicted_labels = torch.sum(predict_levels, dim=1)
    return predicted_labels
